pkg/controller/registry/resolver/querier.go

```diff
@@ -102,24 +106,84 @@ func (q *NamespaceSourceQuerier) FindLatestBundle(pkgName, channelName string, i
 	return nil, nil, fmt.Errorf("%s/%s not found in any available CatalogSource", pkgName, channelName)
 }
 
-func (q *NamespaceSourceQuerier) FindReplacement(bundleName, pkgName, channelName string, initialSource CatalogKey) (*opregistry.Bundle, *CatalogKey, error) {
+func (q *NamespaceSourceQuerier) FindReplacement(currentVersion *semver.Version, bundleName, pkgName, channelName string, initialSource CatalogKey) (*opregistry.Bundle, *CatalogKey, error) {
+	errs := []error{}
+
 	if initialSource.Name != "" && initialSource.Namespace != "" {
 		source, ok := q.sources[initialSource]
 		if !ok {
 			return nil, nil, fmt.Errorf("CatalogSource %s not found", initialSource.Name)
 		}
-		bundle, err := source.GetReplacementBundleInPackageChannel(context.TODO(), bundleName, pkgName, channelName)
+
+		bundle, err := q.findChannelHead(currentVersion, pkgName, channelName, source)
+		if bundle != nil {
+			return bundle, &initialSource, nil
+		}
 		if err != nil {
-			return nil, nil, err
+			errs = append(errs, err)
 		}
-		return bundle, &initialSource, nil
+
+		bundle, err = source.GetReplacementBundleInPackageChannel(context.TODO(), bundleName, pkgName, channelName)
+		if bundle != nil {
+			return bundle, &initialSource, nil
+		}
+		if err != nil {
+			errs = append(errs, err)
+		}
+
+		return nil, nil, errors.NewAggregate(errs)
 	}
 
 	for key, source := range q.sources {
-		bundle, err := source.GetReplacementBundleInPackageChannel(context.TODO(), bundleName, pkgName, channelName)
-		if err == nil {
+		bundle, err := q.findChannelHead(currentVersion, pkgName, channelName, source)
```

Comment:
Since we're iterating over a map, which is non-deterministic, if we have multiple catalogs with the same package/channel and CSVs at head with `olm.skipRange`s that include `currentVersion`, is it possible that we can have non-deterministic upgrades. 

Could we also get false "upgrade available" detections if the range is inclusive of head?